File: tests/performance/drivers/odbc/app/results.cpp
#include "results.h"

#include <sys/utsname.h>

#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>

void write_csv_results(const std::vector<TestResult>& results, const std::string& filename) {
  std::filesystem::path filepath(filename);
  if (filepath.has_parent_path()) {
    std::filesystem::create_directories(filepath.parent_path());
  }

  std::ofstream csv(filename);
  if (!csv.is_open()) {
    std::cerr << "ERROR: Failed to open file for writing: " << filename << "\n";
    return;
  }

  csv << "query_time_s,fetch_time_s\n";

  for (const auto& r : results) {
    csv << std::fixed << std::setprecision(6) << r.query_time_s << "," << r.fetch_time_s << "\n";
  }

  csv.close();
}

std::string get_architecture() {
  struct utsname sys_info;
  if (uname(&sys_info) == 0) {
    std::string machine = sys_info.machine;

    if (machine == "x86_64" || machine == "amd64") {
      return "x86_64";
    } else if (machine == "aarch64" || machine == "arm64") {
      return "arm64";
    }

    return machine;
  }
  return "unknown";
}

std::string get_os_version() {
  const char* os_info = std::getenv("OS_INFO");
  return os_info ? std::string(os_info) : "Linux";
}

void write_run_metadata_json(const std::string& driver_type, const std::string& driver_version,
                             const std::string& server_version, time_t timestamp,
                             const std::string& filename) {
  // Check if metadata file already exists
  std::ifstream check_file(filename);
  if (check_file.good()) {
    check_file.close();
    return;  // Metadata already exists, don't overwrite
  }

  // Detect architecture and OS inside container
  std::string architecture = get_architecture();
  std::string os = get_os_version();

  std::ofstream json(filename);
  if (!json.is_open()) {
    std::cerr << "ERROR: Failed to open metadata file for writing: " << filename << "\n";
    return;
  }

  json << "{\n";
  json << "  \"driver\": \"odbc\",\n";
  json << "  \"driver_type\": \"" << driver_type << "\",\n";
  json << "  \"driver_version\": \"" << driver_version << "\",\n";
  json << "  \"server_version\": \"" << server_version << "\",\n";
  json << "  \"architecture\": \"" << architecture << "\",\n";
  json << "  \"os\": \"" << os << "\",\n";
  json << "  \"run_timestamp\": " << timestamp << "\n";
  json << "}\n";

  json.close();
  std::cout << "✓ Run metadata saved to: " << filename << "\n";
}

File: tests/performance/drivers/core/app/src/results.rs
//! Results output and CSV formatting

use crate::types::IterationResult;
use std::fs;
use std::io::Write;
use std::path::PathBuf;

pub fn write_csv_results(results: &[IterationResult], test_name: &str) -> anyhow::Result<String> {
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();

    // Use RESULTS_DIR env var if set (for local execution), otherwise use /results (Docker)
    let results_dir = std::env::var("RESULTS_DIR").unwrap_or_else(|_| "/results".to_string());
    let results_path = PathBuf::from(&results_dir);
    let filename = results_path.join(format!("{}_core_{}.csv", test_name, timestamp));

    fs::create_dir_all(&results_path)?;
    let mut file = fs::File::create(&filename)?;
    writeln!(file, "query_time_s,fetch_time_s")?;
    for result in results {
        writeln!(
            file,
            "{:.6},{:.6}",
            result.query_time_s, result.fetch_time_s
        )?;
    }

    Ok(filename.display().to_string())
}

pub fn print_statistics(results: &[IterationResult]) {
    if results.is_empty() {
        return;
    }

    let mut query_times: Vec<f64> = results.iter().map(|r| r.query_time_s).collect();
    let mut fetch_times: Vec<f64> = results.iter().map(|r| r.fetch_time_s).collect();

    // Sort for median calculation
    query_times.sort_by(|a, b| a.partial_cmp(b).unwrap());
    fetch_times.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let query_median = if query_times.len() % 2 == 0 {
        (query_times[query_times.len() / 2 - 1] + query_times[query_times.len() / 2]) / 2.0
    } else {
        query_times[query_times.len() / 2]
    };
    let query_min = query_times.iter().cloned().fold(f64::INFINITY, f64::min);
    let query_max = query_times
        .iter()
        .cloned()
        .fold(f64::NEG_INFINITY, f64::max);

    let fetch_median = if fetch_times.len() % 2 == 0 {
        (fetch_times[fetch_times.len() / 2 - 1] + fetch_times[fetch_times.len() / 2]) / 2.0
    } else {
        fetch_times[fetch_times.len() / 2]
    };
    let fetch_min = fetch_times.iter().cloned().fold(f64::INFINITY, f64::min);
    let fetch_max = fetch_times
        .iter()
        .cloned()
        .fold(f64::NEG_INFINITY, f64::max);

    println!("\nSummary:");
    println!(
        "  Query: median={:.3}s  min={:.3}s  max={:.3}s",
        query_median, query_min, query_max
    );
    println!(
        "  Fetch: median={:.3}s  min={:.3}s  max={:.3}s",
        fetch_median, fetch_min, fetch_max
    );
}

fn get_architecture() -> String {
    let arch = std::env::consts::ARCH;

    match arch {
        "x86_64" | "amd64" => "x86_64".to_string(),
        "aarch64" | "arm64" => "arm64".to_string(),
        _ => arch.to_string(), // Return as-is if unknown
    }
}

fn get_os_version() -> String {
    if let Ok(os_info) = std::env::var("OS_INFO") {
        return os_info;
    }
    match std::env::consts::OS {
        "macos" => "MacOS".to_string(),
        "linux" => "Linux".to_string(),
        other => other.to_string(),
    }
}

pub fn write_run_metadata_json(server_version: &str) -> anyhow::Result<String> {
    let results_dir = std::env::var("RESULTS_DIR").unwrap_or_else(|_| "/results".to_string());
    let results_path = PathBuf::from(&results_dir);
    let metadata_filename = results_path.join("run_metadata_core.json");

    // Check if metadata already exists (only write once per run)
    if metadata_filename.exists() {
        return Ok(metadata_filename.display().to_string());
    }

    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();

    // Get driver version from env (set at compile time in Cargo.toml)
    let driver_version = env!("CARGO_PKG_VERSION");

    // Detect architecture and OS inside container
    let architecture = get_architecture();
    let os = get_os_version();

    let metadata = serde_json::json!({
        "driver": "core",
        "driver_type": "universal",
        "driver_version": driver_version,
        "server_version": server_version,
        "architecture": architecture,
        "os": os,
        "run_timestamp": timestamp,
    });

    let mut file = fs::File::create(&metadata_filename)?;
    writeln!(file, "{}", serde_json::to_string_pretty(&metadata)?)?;

    println!("✓ Run metadata saved to: {}", metadata_filename.display());

    Ok(metadata_filename.display().to_string())
}
